consensus/src/dag/dag_store.rs

```diff
@@ -133,14 +88,29 @@ impl Dag {
             .map_or_else(|| highest_round.saturating_sub(1), |_| highest_round)
     }
 
-    pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+    #[cfg(test)]
+    pub fn add_node_for_test(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        self.validate_new_node(&node)?;
+        self.add_validated_node(node)
+    }
+
+    fn add_validated_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
+        let node = Arc::new(node);
+        let round_ref = self
+            .get_node_ref_mut(node.round(), node.author())
+            .expect("must be present");
+        ensure!(round_ref.is_none(), "race during insertion");
+        *round_ref = Some(NodeStatus::Unordered(node.clone()));
+        Ok(())
+    }
+
+    fn validate_new_node(&mut self, node: &CertifiedNode) -> anyhow::Result<()> {
```

Comment:
hmm, it looks a bit weird to have &mut self for validation, but probably it makes insertion simpler

Reply:
yes, it makes insertion and also last validation step simpler.